### PR TITLE
feat(compiler-plugin): add two simple checkers on the frontend part

### PR DESCRIPTION
**Pull request by FreshMag**
Started working on the frontend part of the compiler plugin.

Implemented two checkers:
- `NoAlignInsideALoop`: catches aggregate operations inside a loop without manual `alignedOn`
- `NoAlignOrDealign`: catches explicit usage of `align` and `dealign` inside aggregate functions.

### CODE REVIEW

File: gradle-plugin/src/main/kotlin/it/unibo/.DS_Store
Comment (DanySK) [inline gradle-plugin/src/main/kotlin/it/unibo/.DS_Store:1]: Please remove this one from the tracker

File: settings.gradle.kts
Comment (DanySK) [inline settings.gradle.kts:35]: sort alphabetically and end the file with a newline

File: compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt
Comment (DanySK) [inline compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt:13]: Can you comment this code? It is unclear what it does

File: compiler-plugin-test/src/test/resources/SingleAggregateLine.kt
Comment (DanySK) [inline compiler-plugin-test/src/test/resources/SingleAggregateLine.kt:5]: Can you comment the meaning of %s?

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt
Comment (DanySK) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:11]: `const` in the Companion?

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/MissingAlignExtension.kt
Comment (DanySK) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/MissingAlignExtension.kt:108]: Unless all top level members can be private but one, I would split this into multiple files, possibly creating a new subpackage.

**Review by DanySK — CHANGES_REQUESTED** [review]
reviewed 6/13, left comments to address

File: compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt
Comment (angelacorte) [inline compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt:3]: Can I suggest adding a test where you use the aggregate constructs (neighboring or exchange etc)? In order to have a more complete test.

File: gradle/libs.versions.toml
Comment (angelacorte) [inline gradle/libs.versions.toml:29]: can I suggest to follow the overall semantics? 
for example from this
``` kctfork = "dev.zacsweers.kctfork:core:0.5.1" ```
to this
``` kctfork = { module = "dev.zacsweers.kctfork:core", version = "0.5.1" } ```

Comment (DanySK) [reply gradle/libs.versions.toml:29]: Why? It is just longer and we gain nothing. Unless we use `version.ref`, I'd keep the short version.

The only suggestion here is to use the lexicographic order to simplify the file reading.

**Comment by DanySK** [comment]
@FreshMag you did not allow modifications by maintainers, hence we cannot automatically rebase or merge the mainline into your branch. You need to keep it up to date yourself. For the next PRs, I recommend allowing them.

**Comment by DanySK** [comment]
@FreshMag make sure all conversations are solved before asking for another review round :smile:

File: gradle/libs.versions.toml
Comment (FreshMag) [reply gradle/libs.versions.toml]: I've left the previous format as @DanySK suggested, for now. Also, I've adopted lexicographic order.

File: compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt
Comment (FreshMag) [reply compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt]: Ok, I've added a more flexible unit test in [161237f](https://github.com/Collektive/collektive/pull/454/commits/161237f864d9d37dc4846fd281ce243a8e3d21c1).
I've added a parametric line of code where it is possible to specify the aggregate function to put inside the loop.

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:19]: I prefer tuple-unpacking for more clear code.

```kotlin
.forEach { (functionName, functionCall) -> .... }
```

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestExplicitAlign.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestExplicitAlign.kt:13]: Please, rewrite the test names to be more "flowable" when you read it (nested test assertions).

For instance, I would change as follows:

```kotlin
"The `align` function" - {
  "should produce a warning when used explicitly" {
  ...
}
```

Similarly for the rest of the test assertions.

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:12]: Also here make the test assertions more flowable.

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt:34]: Use markdown syntax to document functions.

https://kotlinlang.org/docs/kotlin-doc.html#inline-markup

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt:37]: Remove `this`

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt:22]:
```suggestion
                "Warning: '$calleeName' method should not be explicitly used",
```

Generally, try to avoid `format` since you have string template

**Review by nicolasfara — CHANGES_REQUESTED** [review]

**Comment by nicolasfara** [comment]
I have two general considerations for this PR:

- The provided tests are not multiplatform, but JVM only (easy fixable I guess)
- From what I can understand, we test that an aggregate function inside a loop without `alignOn` should produce a warning. But what about `listOf(1, 2, 3, 4, 5).forEach { neighboring(it) }`? The same for `map`, `filter`, and so on... at the current stage, do we want to intercept also these cases?

@DanySK @angelacorte

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt
Comment (DanySK) [reply compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:19]: it's called "destructuring declaration" in Kotlin jargon, but I agree.

**Comment by DanySK** [comment]
1. Better fix the MP part before merging. This PR is a kickstart and should build a solid basis for the next steps.
2. I'd go one case at a time. Let's do the for loop first, then we will improve step by step

**Comment by FreshMag** [comment]
Alright, I have changed the requested parts. Have you got any advice on how to fix the MP part?

**Comment by nicolasfara** [comment]
> Does the library you're using support compilation for KMP? Otherwise I think we'll only test the JVM side of the compiler.

Taking a brief look at the dependencies used in this module, for sure
```
apache-commons-codec = "commons-codec:commons-codec:1.17.1"
apache-commons-text = "org.apache.commons:commons-text:1.12.0"
```

those two dependencies are JVM only...

If we want to support KMP we should find an alternative to those libraries.

**Comment by FreshMag** [comment]
The first library is not used, probably a typo by me. I will fix it within the next few hours (by the way, it is also present inside the `compiler-embeddable` project).

The second library is used just for string utilities, but it can be easily replaced by implementing a simple method for string substitution. 

I'm not sure if the [Kotlin compile-testing library](https://github.com/ZacSweers/kotlin-compile-testing) can be a problem as well?

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt
Comment (DanySK) [reply compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt:59]: @FreshMag do you prefer to merge this first and then try to build the DSL (maybe with @nicolasfara's help) or do you want to try it before merging?

**Comment by DanySK** [comment]
@nicolasfara if they're only used in the test project, and if the test library is jvm-only, then it's not a problem is it?

**Comment by nicolasfara** [comment]
> @nicolasfara if they're only used in the test project, and if the test library is jvm-only, then it's not a problem is it?

No, it is not a problem: since this test project is meant to test aggregate anti-pattern or problematic code, it is not mandatory to make it KMP. At the end of the day, at this layer (frontend plugin) we should not deal with platform-specific stuff.

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt:54]: Are all these `println` needed or are they something from your experiments?

**Review by nicolasfara — CHANGES_REQUESTED** [review]

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt
Comment (FreshMag) [reply compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt:54]: I'm removing them right now, sorry. They are not needed.

**Comment by FreshMag** [comment]
I have removed the `apache` libraries that were included inside the testing project. I will leave the Kotlin compile testing library as suggested.

As a first attempt, I have implemented a possible, incomplete DSL to replace the current program template formatting methodology. You can examine it in [this draft](https://github.com/FreshMag/collektive-plugin/pull/6). 
Let me know what do you think about it and please provide possible improvements if needed.

File: compiler-plugin-test/build.gradle.kts
Comment (nicolasfara) [inline compiler-plugin-test/build.gradle.kts:1]: Just realised that this module should not be published in Maven Central, right? @DanySK 

To disable the module's publication the following lines should be added:

```
tasks {
    // Prevent publishing this module since it is a test-only module
    withType<AbstractPublishToMaven>().configureEach {
        enabled = false
    }
    withType<GenerateModuleMetadata>().configureEach {
        enabled = false
    }
}
```

File: compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt:12]: Why do you specify the `pivot` as the input argument?

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt
Comment (nicolasfara) [inline compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt:29]: I'm pretty sure that creating a pivot based on `localId` can be problematic in real code: since all the devices have different IDs, no one would align at the same path (pivot) since it is built based on the ID.

Since here we are testing whether the warning is raised or not, but since the tests act (among the many) as a specification, I would not make the pivot dependent on the `localId`, what do you think @angelacorte @DanySK ?

**Review by nicolasfara — CHANGES_REQUESTED** [review]

**Comment by nicolasfara** [comment]
> I have removed the `apache` libraries that were included inside the testing project. I will leave the Kotlin compile testing library as suggested.
> 
> As a first attempt, I have implemented a possible, incomplete DSL to replace the current program template formatting methodology. You can examine it in [this draft](https://github.com/FreshMag/collektive-plugin/pull/6). Let me know what do you think about it and please provide possible improvements if needed.

I'll take a look in a moment

File: compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt
Comment (nicolasfara) [reply compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt:12]: In that case, the `localId` divides devices based on their ID (even and odd).
In this case, I would avoid having the pivot based on the `localId`

File: compiler-plugin-test/build.gradle.kts
Comment (DanySK) [reply compiler-plugin-test/build.gradle.kts:1]: yeah, makes sense

File: compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt
Comment (DanySK) [reply compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt:29]: For testing purposes, I don't really see the problem

Comment (nicolasfara) [reply compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt:29]: I would have preferred a static string instead, but that's okay.
Mark as resolved.

File: compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt
Comment (FreshMag) [reply compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt:12]: Ok, will fix it in a moment.

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt:22]:
```suggestion
                "Warning: '$calleeName' method should not be explicitly used",
```

**Review by nicolasfara — APPROVED** [review]

**Review by angelacorte — APPROVED** [review]

**Review by DanySK — APPROVED** [review]

**Pull request merged**
